core/src/main/scala/kafka/tools/EndToEndLatency.scala

```diff
@@ -58,35 +69,57 @@ object EndToEndLatency {
     producerProps.put(ProducerConfig.ACKS_CONFIG, producerAcks.toString)
     producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
     producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
-    val producer = new KafkaProducer[Array[Byte],Array[Byte]](producerProps)
+    val producer = new KafkaProducer[Array[Byte], Array[Byte]](producerProps)
+
+    def finalise() {
+      consumer.commit(CommitType.SYNC)
+      producer.close()
+      consumer.close()
+    }
 
-    // make sure the consumer fetcher has started before sending data since otherwise
-    // the consumption from the tail will skip the first message and hence be blocked
-    Thread.sleep(5000)
+    //Ensure we are at latest offset
+    var recordIter = consumer.poll(0).iterator
+    consumer.seekToEnd()
 
-    val message = "hello there beautiful".getBytes
     var totalTime = 0.0
     val latencies = new Array[Long](numMessages)
+
     for (i <- 0 until numMessages) {
+      val message = arrayOfLen(messageLen)
       val begin = System.nanoTime
-      producer.send(new ProducerRecord[Array[Byte],Array[Byte]](topic, message))
-      val received = iter.next
+
+      producer.send(new ProducerRecord[Array[Byte], Array[Byte]](topic, message))
```

Comment:
I've added a check to ensure producer acks are either "1" or "all". As I understand it this makes send() a blocking call so long as LINGER_MS_CONFIG = 0. 

The call to poll() should now block correctly post KAFKA-2486. 

There is an alternative implementation where a callback can be passed to send() and used to trigger the consumer.poll(). I didn't go this way just because it seemed unnecessary. 

If you my understanding is wrong - happy to change